core/solver/multigrid.cpp

```diff
@@ -162,336 +162,384 @@ void handle_list(
+                            e.get(), false, false, true);
+        } else if ((cycle == multigrid::cycle::kfcg ||
+                    cycle == multigrid::cycle::kgcr) &&
+                   level % multigrid->get_parameters().kcycle_base == 0) {
+            // kcycle_state.kstep<VT>(cycle, level, g, e);
+            GKO_NOT_SUPPORTED(cycle);
         }
+    }
+    // prolong
+    mg_level->get_prolong_op()->apply(next_one, e.get(), next_one, x_ptr);
+
+    // end or origin previous
+    bool use_post = is_end || mid_case == multigrid::mid_smooth_type::both ||
+                    mid_case == multigrid::mid_smooth_type::post_smoother;
+    // post-smooth
+    if (use_post && post_smoother) {
+        post_smoother->apply(b_ptr, x_ptr);
+    }
 
-        // put the mid smoother into the end of previous cycle
-        // only W/F cycle
-        bool use_mid =
-            (cycle == multigrid::cycle::w || cycle == multigrid::cycle::f) &&
-            !is_end && mid_case == multigrid::mid_smooth_type::standalone;
-        if (use_mid && mid_smoother) {
-            mid_smoother->apply(b, x);
-        }
+    // put the mid smoother into the end of previous cycle
+    // only W/F cycle
+    bool use_mid =
+        (cycle == multigrid::cycle::w || cycle == multigrid::cycle::f) &&
+        !is_end && mid_case == multigrid::mid_smooth_type::standalone;
+    if (use_mid && mid_smoother) {
+        mid_smoother->apply(b_ptr, x_ptr);
     }
+}
 
-    struct KCycleMultiGridState {
-        void reserve_space(MultigridState* mg_state_in)
-        {
-            mg_state = mg_state_in;
-            auto k_num = (mg_state->multigrid->get_mg_level_list().size() - 1) /
-                         mg_state->multigrid->get_parameters().kcycle_base;
-            alpha_list.reserve(k_num);
-            beta_list.reserve(k_num);
-            gamma_list.reserve(k_num);
-            rho_list.reserve(k_num);
-            zeta_list.reserve(k_num);
-            v_list.reserve(k_num);
-            w_list.reserve(k_num);
-            d_list.reserve(k_num);
-            old_norm_list.reserve(k_num);
-            new_norm_list.reserve(k_num);
-        }
+void MultigridState::KCycleMultiGridState::reserve_space(
+    MultigridState* mg_state_in)
+{
+    mg_state = mg_state_in;
+    auto k_num = (mg_state->multigrid->get_mg_level_list().size() - 1) /
+                 mg_state->multigrid->get_parameters().kcycle_base;
+    alpha_list.clear();
+    beta_list.clear();
+    gamma_list.clear();
+    rho_list.clear();
+    zeta_list.clear();
+    v_list.clear();
+    w_list.clear();
+    d_list.clear();
+    old_norm_list.clear();
+    new_norm_list.clear();
+    alpha_list.reserve(k_num);
+    beta_list.reserve(k_num);
+    gamma_list.reserve(k_num);
+    rho_list.reserve(k_num);
+    zeta_list.reserve(k_num);
+    v_list.reserve(k_num);
+    w_list.reserve(k_num);
+    d_list.reserve(k_num);
+    old_norm_list.reserve(k_num);
+    new_norm_list.reserve(k_num);
```

Comment:
Maybe extract into a function that does both ? Something like this (untested code)

```c++
namespace detail {

template <typename T>
void clear_and_reserve(T* obj, int size) {
obj->clear();
obj->reserve(size);
}
```